text/0036-auto-instrumentation-ui-thread.md

```diff
@@ -0,0 +1,77 @@
+
+This attribute is also valid for any Frontend SDK.
+
+# Background
+
+Mobile and Desktop developers must be very aware of the UI Thread(s). It's a core part of developing 
+user interfaces and a common source of bad user experience due to unresponsive UI.
+
+File I/O operations on main thread is one of [the common causes](https://developer.android.com/topic/performance/vitals/anr#io-on-main) 
+for slowing down mobile applications and worsening User Experience. They can result in dropped frames, or even worse, 
+in App Hangs/ANR, which can happen if the UI/Main thread is blocked for more than 5 seconds. It is also easy to miss 
+those operations, as they can be hidden under some third-party libraries/SDKs (e.g. image loading library, which does caching).
+
+Since we already offer auto-instrumentation for File I/O operations on mobile platforms (Android, iOS, Dart),
+it will be an easy win for us to start detecting Performance Issues by introducing a relatively small change
+to the `Span.data` map.
+
+[ANR documentation as part of Android Vitals](https://developer.android.com/topic/performance/vitals/anr).
+
+# Options Considered
+
+## Detecting UI/Main thread for all Spans (Option 1)
+
+The approach was to detect whether *any* span (including custom ones) has run on the UI/Main thread or not.
+However, this turned out to be problematic, because we can't certainly say how much time a span spent on the
+UI/Main thread vs other threads. The span might be started on the UI/Main thread, then perform a network
+request on a background thread (a new child span), and then finished back on the UI/Main thread. It's unclear
+whether this span should be marked as run on UI/Main thread or not.
+
+This option would also require quite some changes in the protocol (e.g. having a map of threads with the start/end timestamp pairs)
+as well as frontend changes.
+
+This option has been decided against, due to potential high-complexity of the proper solution or just simply
+infeasiable on some certain platforms.
+
+## Making `ui_thread` attribute part of the protocol (Option 2)
+
+Since this flag is only relevant to Frontent SDKs, and is only used to mark auto-instrumented spans, it was
+decided against adding it to the common `Span` interface of the protocol.
+
+# Drawbacks
+
+The only drawback we found with this approach is that it works only for automatically created spans and
+cannot be expanded to custom spans. However, given, that our auto-instrumentation offering is growing, this 
+should not be a problem. Other potential detectors we could create for detecting performance issues:
+
+* Network I/O on main thread
+* Database operations on main thread
+* Resources (images/audio/video) manipulation on main thread (e.g. decoding, resizing, etc.)
+
+# Unresolved questions
+
+* Should the `ui_thread` attribute be part of the protocol or just a new key in the data bag?
```

Comment:
I guess this is the main question here to you folks @AbhiPrasad and @jjbayer (unless we say this option is total garbage).

Reply:
I really like this direction, and I think we can make it a forwards compatible data bag. This means we treat certain keys as special cased in Relay, but we don't need to restrict the data bag to just those values.

These keys have special _semantic meaning_ - this is how OpenTelemetry operates as well. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/README.md

We can maintain a list of these keys in our develop docs/relay - so we can use them across SDKs.

Reply:
It’s because this can potentially balloon as we want to do similar analysis on other span attributes.

I think of the data bag on the span similar to contexts on events - we can special case certain values in there to treat them differently during processing/in the product.

Reply:
The problem is that the `data` bag is open for users as well.
One can easily coincidentally use the very same key and break the feature.
Contexts on the event have the same problem.
A few alternatives.
* Top level - optional field
* A Sentry prefix for special cased keys in the `data` bag (avoiding collision)
* A different `data` bag only for special cased keys

Reply:
This same thing applies for contexts as well, as well as span op and description (which is also used by detectors).

the likelyhood of this happening in my opinion is very low since

a) we are setting on data not tags (set tag is probably more used for spans since it matches the set tag for events)
b) spans are not indexed, so most people are probably setting attributes on the transaction, not the span.

Reply:
Less likely than contexts, yes, but this problem exists for contexts, so maybe a `prefix` won't hurt.

Reply:
> maybe a prefix won't hurt

My hesitation against a prefix here is that we'll end up adding more and more of these, and I'd rather we just start a set of semantic conventions, and then educate our users about this - it's a more scalable approach.

What is the chance that someone is going to grab the active span and set this exact data field?

Reply:
I can't tell you the chance but this is a known issue for Contexts, customers already came up with "bugs" because they were using "our reserved" words, people don't look up the docs for that, so, it will happen.

Reply:
@marandaneto @AbhiPrasad I'm thinking about changing the attribute name  to something more explicit - for example `blocked_ui_thread`. As @mattjohnsonpint mentioned during the TSC, `ui_thread` might be too misleading to be applied for _any_ auto-instrumentation, because e.g. for activity/viewcontroller transactions (that are also auto-instrumented) we cannot say for sure how much time they spent on the UI vs background thread.

`blocked_ui_thread` would be more explicit in telling that the span was actually blocking the UI thread its entire duration.

What do y'all think?

Reply:
A TR/Span can technically start in the UI thread and finish in the background, and vice-versa, are we considering this use case?

Reply:
not with this RFC, this is solely focused on auto-instrumented spans, that cannot span across multiple threads, like File I/O

Reply:
I see, the problem is that the naming is very generic, but it has a specific scope (such as auto instrumentation for File IO), so I'd either put a more specific naming or the RFC should also take this under consideration, otherwise this will be misused or be open to misinterpretation.

Reply:
I agree, that's why I think `blocked_ui_thread` is more specific in telling that the span has blocked the UI thread for its entire duration. Or do you think it's still too generic?

We could also change it to something like `io_ui_thread` which would also work for Network I/O or Database I/O

Reply:
IMO `blocked_ui_thread` is a good option

Reply:
@AbhiPrasad we are already using the `data` field without prefixes so I am fine with it, even though they are "reserved" by sentry such as [http.status_code](https://github.com/getsentry/develop/issues/445), so, I agree.

@romtsn fine with the `blocked_x_thread` as well, but maybe we use `main` instead of `ui`? the same would apply for a CLI app that has its main thread blocked but its not an UI app.

Reply:
I'm fine with `main`, but `ui` sounds like it needs more attention (in the mobile/desktop world).

Also, if the main thread is blocked in a CLI app it's.. probably fine(?), since there's no user interface involved and maybe it's intentional

Reply:
The blocked `ui` thread can also be intentional, I'm just trying to make it more generic so it's applicable for other platforms, your call here.
https://developer.android.com/guide/components/processes-and-threads mentions many times the name `main`, also if you get the current `Looper` on Android, its called `main`, so its semantically correct.

Reply:
I'm good with `main`, but would need @wmak to change it on the detector's site then

Reply:
Feel free to change it to main :+1: I can update the detectors. Just let me know what the final decision is

Reply:
alright, I changed it to `blocked_main_thread` on Android and will update the RFC as well. cc @wmak @brustolin